### DISCUSSION TITLE
Architecture Options

### DISCUSSION THREAD

**Discussion by docuracy** [WorldHistoricalGazetteer/place#23] [General]
The World Historical Gazetteer (WHG) currently relies on two DigitalOcean servers, both located in England. One server handles core data storage and retrieval through a traditional relational database, with Elasticsearch providing search and indexing capabilities. The second server is dedicated to generating and serving maps. While the Elasticsearch indexes have supported WHG in its early development phases, they are now insufficient for the growing demand for more advanced search features, including phonetic and semantic searching, matching, and match-ranking.

To meet these evolving needs, we propose migrating much of the data storage and indexing to a combined vector database and indexing system. Vespa is an ideal solution because it supports both vector search and machine learning models, providing a more powerful, flexible, and scalable system for handling our advanced search requirements. Vespa, like the rest of WHG’s software stack, is open source and free of licensing costs. However, Vespa’s full capabilities, particularly its use of machine learning models for phonetic and semantic matching and ranking, require enhanced server infrastructure, specifically GPUs (Graphics Processing Units) in addition to CPUs (Central Processing Units).

Vespa would facilitate the development of the **WHG PLACE** (Place Linkage, Alignment, and Concordance Engine), a system designed not only to improve search results delivered through our APIs and on our website, but also to significantly streamline the incorporation of newly contributed datasets into WHG. The PLACE engine would greatly reduce the manual effort required to reconcile, match, and rank places across different datasets—an essential step in the ongoing establishment of WHG as a central resource in the Digital Humanities ecosystem.

We have been offered the use of a dedicated, high-performance server by the MEHDIE research group at the University of Aalborg, Denmark. This server, available from the end of 2024 at no cost for several years, would provide the necessary GPU support and computational power to fully leverage Vespa’s capabilities. While this would eliminate much of WHG’s reliance on commercial servers, it would still be beneficial to maintain a minimal, secondary server—likely hosted on DigitalOcean—for the purposes of backup, load balancing, and distributing query and indexing operations.

| **Option**                                  | **1: Enlarging DigitalOcean Server**                                                                                                                                                                  | **2: Use Donated Standalone Server with GPU Capability**                                                                                                                                  |
|---------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| **Description**                             | Continue hosting the application entirely on DigitalOcean, scaling up resources to accommodate Vespa. Merge the split-server architecture into a single, larger server, with backup in cold cloud storage. No hosted GPU support means vector embeddings for indexes would need to be generated offline, with live queries running on CPU.                                                                                                                                 | Host WHG on the Aalborg server with GPU support. Use DigitalOcean only for replication, backup, and load balancing. |
| **Pros/Cons**                                    | - Cost increase: Significantly higher server costs to scale up resources for Vespa, especially for CPU-bound vector operations.<br>- Limited performance: Offline embedding generation adds complexity, and live queries on CPU would be slower compared to GPUs.<br>- Storage: Maintaining backups will entail additional complexity and costs.<br>- Dependency: Reliance on external institutional support, which could pose risks if the arrangement ends or changes. | - Independence from external institutional support.<br>- Cost savings: Very significant reduction in server costs due to free hosting.<br>- GPU availability: Access to GPU resources for fast generation of vector embeddings for both indexing and live queries.<br>- Performance: Faster live query processing with GPUs, especially for machine learning models.<br>- Storage capacity more than adequate to extend the range of map features we can host, including DEM (Digital Elevation Model) and OSM (OpenStreetMap) vector tiles. |
| **Server Configuration and Likely Costs**  | - Minimum DigitalOcean Premium AMD (8 vCPUs, 32GB RAM, 320GB Storage, 10TB Network): $168/month<br>- Cold Storage: 42GB (AWS Glacier based on current database and indexes, compressed and with 7 versioned backups) $0.04/month<br>- **Summary**: 156% of current | - DigitalOcean Regular (2 CPUs, 4GB RAM, 80GB Storage, 4TB Network): $24/month<br>- Lenovo SR635v3 AMD Epyc 9354p (32 CPUs + 2x Nvidia L4 24Gb GPUs, 768GB RAM (12x64GB DDR5), 2x480GB SATA SSD TIL system + 1x 3.8TB NVMe Storage): $free<br>- **Summary**: 22% of current |

### Summary

The decision to migrate WHG’s indexing and search capabilities to Vespa is driven by the need for more sophisticated search features and the growing volume of WHG's data. Vespa’s open-source nature and advanced features make it a suitable candidate for this migration.

Both hosting options have their merits:

- **Option 1 (Enlarging DigitalOcean Server):** Offers independence from external institutional support, but comes at a higher cost and lacks GPU support, limiting performance.
- **Option 2 (Standalone Server with GPU Capability):** Provides a cost-effective, high-performance solution with GPU resources, but introduces dependency on external support.

Given the substantial cost savings and performance benefits, Option 2—using the donated standalone server with GPU support—seems the most viable option, provided that potential risks associated with institutional dependency are mitigated through proper agreements and fallback plans.

Further details are provided in the GitHub Issues at:
https://github.com/WorldHistoricalGazetteer/whg3/issues/381
https://github.com/WorldHistoricalGazetteer/whg3/issues/372

**Comment by docuracy** [comment]
## Initial Questions

### 1. **Scalability & Performance:**

>    - How do you assess the scalability of each infrastructure option (DigitalOcean vs. Aalborg server with GPU)? Are there any potential bottlenecks that we should anticipate?
>    - Do you think the suggested GPU support for indexing and live queries will provide the anticipated performance improvement?
>    - Should we consider other alternatives, such as distributed computing or hybrid cloud setups, to meet our future scalability needs?

### 2. **Cost and Budget Considerations:**

>    - What are your thoughts on the proposed costs for upgrading the DigitalOcean server vs. utilising the donated server? Are there any hidden costs we should be aware of (e.g., maintenance, technical debt)?
>    - Do you think that free hosting with GPU capabilities will provide enough long-term sustainability, considering potential reliance on external institutional support? How do we mitigate the risk of this dependency?

### 3. **Technical Feasibility:**

>    - Given the proposed split architecture (using DigitalOcean for replication and backup, with the GPU server hosting WHG), how do you see this impacting our data management and synchronisation between servers?
>    - Are there any potential issues with integrating Vespa into the DigitalOcean infrastructure if we scale up resources? Are there specific configurations that could optimize this setup?
>    - Do you foresee any challenges in setting up or maintaining the donated server, particularly with the GPU resources and storage capabilities?
>    - Would additional tools or systems (e.g., monitoring tools, scaling automation) might be necessary to ensure the chosen infrastructure remains performant and reliable?

### 4. **Data Management & Backup:**

>    - Are the current plans for cold storage and backup using AWS Glacier sufficient, or should we explore other backup strategies (e.g., on-premises backup or cloud-based backup services with higher availability)?
>    - Should we consider replication across multiple cloud providers or geographies for disaster recovery purposes? How critical is this for our current use case?

### 5. **Security and Compliance:**

>    - What security considerations should we take into account when scaling the infrastructure, particularly for cloud-based and external servers?
>    - Do we need additional encryption or access control measures?
>    - Are there any regulatory or compliance requirements that need to be considered when hosting the data or scaling infrastructure in different regions or platforms?

### 6. **Future-Proofing:**

>    - As the World Historical Gazetteer grows, how can we ensure that the infrastructure upgrade is flexible enough to accommodate future needs, such as additional datasets and AI/ML capabilities?
>    - What steps should we take to ensure that the system is capable of scaling with new features like mBERT embeddings, and future machine learning models, as well as handling potentially more complex geospatial queries?
>    - How do you envision the infrastructure evolving—if at all—over the next 3-5 years?

### 7. **Dependencies and Risk Mitigation:**

>    - What risks do you foresee in relying on a donated server (particularly in terms of maintenance, uptime, and support)? How can we mitigate these risks?
>    - How should we approach potential downtimes, data loss, or failures with either option? What contingency plans (e.g., backups, failover strategies) would you recommend?

### 8. **Collaboration and Communication:**

>    - How can we ensure effective communication and collaboration among the team members working on infrastructure management, development, and scaling? Are there tools or workflows that can help us with this?
>    - Are there specific goals or milestones we should set to track progress and ensure the infrastructure meets the needs of our users and stakeholders?

### 9. **Long-Term Strategy:**

>    - How should we structure our infrastructure to allow for future collaborations with external partners or researchers, particularly when scaling data storage, access, and computational resources?
>    - What would be your recommendation for ongoing infrastructure maintenance and updates to ensure the system remains relevant, secure, and performant over time?

### 10. **Open-Ended:**

>    - What other considerations or suggestions do you have that haven’t been covered in the document regarding scaling the infrastructure or improving performance?
>    - How do you think the new infrastructure will align with the broader strategic goals of the World Historical Gazetteer?

**Comment by tomersagi** [comment]
The donated server will be in a well-maintained server-farm with good SLA. I have good experience with AAU IT services. That being said - it introduces a single point of failure and servers can fail over time. Especially their hard drives. We should aim for docker-based one click deployability to allow deploying a backup server if the main one fails. A disaster recovery (DR) plan should be in place and rehearsed occasionally and a monitoring system alerting us in case of a server / website outage should be implemented.

**Reply by docuracy** [reply]
Good point. That's what I'm working on now with a Kubernetes setup.

**Reply by tomersagi** [reply]
By the way, the AAU setup includes two servers - the GPU server for the vector-database, querying, inference, matching tasks, etc, and a web-server for serving the application and it's app-database (for users, datasets, collections, etc. ). The latter server is already installed and ready for us.

**Reply by docuracy** [reply]
Sounds great!

For those who don't know, usually one server in a Kubernetes cluster would act as a "control-plane" to orchestrate other servers acting as "worker" nodes. In our AAU set-up I would suggest ~that the GPU server and _perhaps_ additionally a DigitalOcean server would be worker nodes, and the other AAU server would be configured as both control-plane _and_ worker. The DigitalOcean server might also then act as a standby control-plane+worker to provide basic service in the event of outage or maintenance of the AAU machines.~ see below...

**Comment by docuracy** [comment]
### Service Placement suggestion

This distribution would need some adjustment to ensure that the DigitalOcean server is capable of providing effective disaster recovery.

| Service                  | Denmark Server 1 | Denmark Server 2 (GPU) | DigitalOcean       |
|--------------------------|------------------|-------------------------|--------------------|
| Kubernetes Control Plane | ✅               | ❌                      | ❌                 |
| Django App               | ✅               | Secondary               | ❌                 |
| PostgreSQL               | Primary Replica  | ❌                      | Secondary Replica  |
| pgBackRest               | ✅                       | ❌                      | Filespace or Object Storage   |
| Redis + Celery           | ✅               | Secondary               | Redundant Backup   |
| Contour (Ingress)        | ✅               | ❌                      | ❌                 |
| Flannel                  | ✅               | ✅                      | ✅                 |
| MetalLB                  | ✅               | ❌                      | ✅                 |
| Node.js Services         | Secondary        | ✅                      | Scaling Needs Only |
| Tippecanoe               | ❌               | ✅                      | ❌                 |
| Vespa (Controller)       | ✅               | ❌                      | ❌                 |
| Vespa (Worker 1)         | ❌               | ✅                      | ❌                 |
| Vespa (Worker 2)         | ❌               | ❌                      | ✅                 |
| Monitoring               | ✅ (Prometheus + Grafana) | ❌            | ❌                 |
